src/main/java/org/assertj/core/api/AbstractByteBufferAssert.java

```diff
@@ -0,0 +1,498 @@
+   *
+   * // ... this one also fails but because the charsets do not match.
+   * assertThat(buffer).endsWith("test", StandardCharsets.UTF_16);</code></pre>
+   *
+   * @return {@code this} assertion object.
+   * @throws AssertionError if the actual {@code ByteBuffer} is {@code null}.
+   * @throws AssertionError if the actual {@code ByteBuffer} is empty.
+   * @throws AssertionError if the actual {@code ByteBuffer}s content does not end with the given expected string.
+   */
+  public SELF endsWith(String expected, Charset charset) {
+    isNotEmpty();
+
+    String contentString = new String(getContent(actual), charset);
+    if (!contentString.endsWith(expected)) throwAssertionError(contentsShouldEndWith(expected, actual, charset));
+    return myself;
+  }
+
+  /**
+   * Verifies that the content of the actual {@code ByteBuffer} ends with the given expected byte array.
+   *
+   * Example:
+   *
+   * <pre><code class='java'> ByteBuffer buffer = ByteBuffer.wrap("test".getBytes());
+   *
+   * // this assertion succeeds ...
+   * assertThat(buffer).endsWith("test".getBytes());
+   *
+   * // ... but this one fails as the content of "buffer" does not end with the given expected byte array.
+   * assertThat(buffer).endsWith("test two".getBytes());</code></pre>
+   *
+   * @return {@code this} assertion object.
+   * @throws AssertionError if the actual {@code ByteBuffer} is {@code null}.
+   * @throws AssertionError if the actual {@code ByteBuffer} is empty.
+   * @throws AssertionError if the actual {@code ByteBuffer}s content does not end with the given expected byte array.
+   */
+  public SELF endsWith(byte[] expected) {
+    isNotEmpty();
+
+    byteArrays.assertEndsWith(info, getContent(actual), expected);
+    return myself;
+  }
+
+  /**
+   * Verifies that the content of the actual {@code ByteBuffer} ends with the given expected {@code ByteBuffer}.
+   *
+   * Example:
+   *
+   * <pre><code class='java'> ByteBuffer actual = ByteBuffer.wrap("test".getBytes());
+   *
+   * // this assertion succeeds ...
+   * ByteBuffer expected = ByteBuffer.wrap("test".getBytes());
+   * assertThat(actual).endsWith(expected);
+   *
+   * // ... but this one fails as the content of "buffer" does not end with the given expected {@code ByteBuffer}.
+   * ByteBuffer expected = ByteBuffer.wrap("test two".getBytes());
+   * assertThat(actual).endsWith(expected);</code></pre>
+   *
+   * @return {@code this} assertion object.
+   * @throws AssertionError if the actual {@code ByteBuffer} is {@code null}.
+   * @throws AssertionError if the actual {@code ByteBuffer} is empty.
+   * @throws AssertionError if the actual {@code ByteBuffer}s content does not end with the given expected {@code ByeBuffer}.
+   */
+  public SELF endsWith(ByteBuffer expected) {
+    isNotEmpty();
+
+    byteArrays.assertEndsWith(info, getContent(actual), getContent(expected));
+    return myself;
+  }
+
+  private byte[] getContent(ByteBuffer buffer) {
```

Comment:
Good idea, are there different implementations/subclasses of byte buffer that would behave differently if you duplicate not as a subclass but as a ByteBuffer ? 

I'm asking because we did a similar thing with maps for some assertions and duplicating a specific implementation of map that accepted string keys ignoring their case, we ended with a duplicated hash map that did not ignore the case.

Overall, we can duplicate the byte buffer and see later if we have some corner cases because of it,  we should make it clear in the javadoc that we are duplicating the buffer.

Reply:
My point is that Buffer#duplicate() creates a shallow copy of the buffer with shared data under the hood, so it's safe to manipulate the buffer's pointers (position, limit, mark) without the need to restore it after modification.

Regarding ByteBuffer subclasses, there is only one - MappedByteBuffer - which behaves in the same manner as the rest.

Reply:
I am fine with this change, but I am not sure I understand the benefit. Ultimately, it seems both approaches leave the buffer in the same state as it originally was, so what is the benefit of doing the shallow copy?

Reply:
IMHO, it is a good approach to work on a copy of the original buffer in all places by default - it's extremely easy to change the buffer's pointers, which would fail following assertions.
You have the right in this case - the result would be the same with both implementations.

Reply:
I agree that it is quite easy to make changes to the buffer's pointers, so I am making the suggest change. 
I noticed however that the method you suggested above uses `copy.remaining()` instead of `copy.limit()` for the size of the `content` array.

I think in this case, it should be `copy.limit()` to prevent errors in cases where the buffer's position has already moved, as otherwise the entire content of the buffer will not fit in the `content` array.

But it did get me thinking: all the assertions I defined here are under the assumption the want to assert on the contents of the _entire_ buffer. So **not** relative to the current position. So for example, assuming I have a buffer with the String `test` in it, and my position is already on the `s`. Then doing `assertThat(buffer).endsWith("est")` would pass. Whereas if I did assertions relative to the position then `assertThat(buffer).endsWith("est")` would fail. But `assertThat(buffer).endsWith("st")` would succeed for both. 

Personally I think doing assertions on the content of the entire buffer makes sense. But I don't use buffers that often, so I don't want make any assumptions as to what others might find most useful.